Switch Vert.x, datasource, reactive datasource and Agroal to @ConfigMapping

gsmet opened the pull request:
/cc @geoand @radcortez @yrodiere

gsmet committed:
Switch to @ConfigMapping for Vert.x extension

gsmet committed:
Convert datasource and reactive datasource config to @ConfigMapping

gsmet commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:
```diff
@@ -248,9 +247,11 @@ public static void startServerAfterFailedStart() {
-            VertxConfiguration vertxConfiguration = new VertxConfiguration();
-            ConfigInstantiator.handleObject(vertxConfiguration);
-            vertx = VertxCoreRecorder.recoverFailedStart(vertxConfiguration).get();
+            // TODO: we can't do this anymore
+            //VertxConfiguration vertxConfiguration = new VertxConfiguration();
+            //ConfigInstantiator.handleObject(vertxConfiguration);
```
I don't really know how I can do this. AFAICS, it parses specifically the Vert.x HTTP config to be able to start a Vert.x HTTP server without triggering other errors if possible.
This is a big hack, I'm unsure we want to keep this behavior? If so, I think it's important for @radcortez to have a look anyway.

gsmet replied:
Yes, the idea was that we would still try to get a minimal dev mode up even if something is very wrong. It is useful for sure and we need to keep this feature. I commented it out so that I was able to compile the extension and get the tests running.

radcortez replied:
Well, this still relies on `SmallRyeConfig` to populate the object, so if you have an invalid configuration it would still fail:

```
Caused by: java.lang.RuntimeException: java.lang.IllegalArgumentException: SRCFG00039: The config property quarkus.http.port with the config value "x" threw an Exception whilst being converted SRCFG00029: Expected an integer value, got "x"
	at io.quarkus.runtime.configuration.ConfigInstantiator.handleObject(ConfigInstantiator.java:139)
	at io.quarkus.runtime.configuration.ConfigInstantiator.handleObject(ConfigInstantiator.java:78)
	at io.quarkus.runtime.configuration.ConfigInstantiator.handleObject(ConfigInstantiator.java:57)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder.startServerAfterFailedStart(VertxHttpRecorder.java:264)
	... 9 more
```

Unless I'm missing something doing:

```java
SmallRyeConfig config = ConfigProvider.getConfig().unwrap(SmallRyeConfig.class);
VertxConfiguration vertxConfiguration = config.getConfigMapping(VertxConfiguration.class);
```

Should be enough.

radcortez replied:
This is now supported automatically with https://github.com/smallrye/smallrye-config/pull/939

radcortez commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSourcesJdbcBuildTimeConfig.java:
```diff
@@ -5,35 +5,35 @@
      */
-    @ConfigItem
-    public DataSourceJdbcBuildTimeConfig jdbc;
+    DataSourceJdbcBuildTimeConfig jdbc();
 
     /**
      * Additional named datasources.
      */
     @ConfigDocSection
     @ConfigDocMapKey("datasource-name")
-    @ConfigItem(name = ConfigItem.PARENT)
-    public Map<String, DataSourceJdbcOuterNamedBuildTimeConfig> namedDataSources;
+    @WithParentName
+    Map<String, DataSourceJdbcOuterNamedBuildTimeConfig> namedDataSources();
```
The idea was to add support for default values in a Map. I started some prototypes and experimented with some ideas,  and I think I had a plan to implement that, but it is not done yet.

Adding a specific class to support this, while possible, of course, would fall off the traditional way of mapping stuff. Until now, we were able to use standard types to represent pieces of the configuration. Adding specific constructs will always require a lot of adoption effort (and probably some frustration on why can't you use what you already have).

What will not be inline with a regular `Map`? I'm thinking that a default will not be listed in `.keySet()`, `entrySet()`, and `containsKey()`, but a `get` will return you a default if there is no value for that key.

radcortez replied:
Also, unofficially you can try this to initialize a group programmatically. I haven't tried that with complex examples, but it should work fine, as long as every mapping has a default defined or you get an error:

```java
SmallRyeConfig config = ConfigProvider.getConfig().unwrap(SmallRyeConfig.class);
ConfigMappingContext context = new ConfigMappingContext(config);
// you need to add the path to the map, plus `*` to represent the key. This will not be added to the main mapping `Map`
context.getStringBuilder().replace(0, context.getStringBuilder().length(), "map.nested.*");
ConfigGroup group = context.constructGroup(ConfigGroup.class);
```

yrodiere replied:
> What will not be inline with a regular Map?

> What will not be inline with a regular `Map`? I'm thinking that a default will not be listed in `.keySet()`, `entrySet()`, and `containsKey()`, but a `get` will return you a default if there is no value for that key.

Yes, that.

Also, and perhaps more importantly, unless you make it configurable through annotations this is a breaking change. There's a decent chance some code out there (probably even in Quarkus) was using `map.get("something") == null` to test whether something was configured or not. Yes I know there is `containsKey`, but `var val = config.get(...); if ( val == null ) {...} else {...}` is a common pattern when you need to use the value if it's there.

By hijacking `Map` to support more than its contract, we would break such existing code.

> Adding specific constructs will always require a lot of adoption effort

Note we're not requesting that anyone use that theoretical new type. It's not a replacement for existing features, it comes in addition to existing `Map` support. I'm not sure how adoption effort would be relevant in that case?

> (and probably some frustration on why can't you use what you already have).

And implementing behavior that does not match the `Map` contract (see above) will lead to nasty bugs, which is *also* frustrating.

Even if the behavior is configurable through annotations, I wouldn't be surprise to see it bite someone who comes back to the code after a while and just expects a `Map` to behave like a `Map`.

> Also, unofficially you can try this to initialize a group programmatically

Nice trick, maybe we could just add the feature to Quarkus then? Then we have fewer constraints as to how it should be exposed (custom type vs. `Map`).

gsmet committed:
Switch Redis extension to @ConfigMapping

gsmet committed:
Reorganize Redis deployment package

This unifies the config reference doc and make sure we have both runtime
and build time config in the same file.

gsmet commented:
@radcortez I have been struggling with this for a while and I think I will need your help.
I had to convert the Redis extension because it is using some config classes from Vert.x.

Unfortunately, I have  two issues converting this extension:

- `NetConfig` has an `Optional<ProxyConfig> proxyOptions()` configuration property. It used to be an empty `Optional` when no proxy configuration was set, it is now defined (with the default values set). This causes problems as it enables the proxy configuration. I think this new behavior will cause some problems for a lot of extensions so I suppose we should try to go back to the existing behavior. Or we will have to find another way to cover that use case.
- I wonder if `@WithConverter` works similarly to `@ConvertWith` when it comes to `List` because I also have this error:
```
[ERROR]   NonDefaultFileForDefaultClientImportPreloadingTest » Runtime java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.redis.deployment.client.RedisClientProcessor#init threw an exception: java.lang.ClassCastException: class java.lang.String cannot be cast to class java.util.List (java.lang.String and java.util.List are in module java.base of loader 'bootstrap')
	at io.quarkus.redis.deployment.client.RedisClientProcessor.getRedisLoadScript(RedisClientProcessor.java:300)
```

I added some comments in the code, I will point them to you in subsequent GitHub comments.

I will start to test your Map default branch now and report back.

gsmet commented on extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/client/VertxRedisClientFactory.java:
```diff
@@ -105,61 +104,62 @@ private static void customize(String name, RedisOptions options) {
+
+        tcp.localAddress().ifPresent(net::setLocalAddress);
+        tcp.nonProxyHosts().ifPresent(net::setNonProxyHosts);
+        // TODO @radcortez proxyOptions() used to be empty when nothing was set and is not empty anymore
```
This is where the problem with `Optional` manifests itself.

radcortez commented:
> @radcortez I have been struggling with this for a while and I think I will need your help.

I'll have a look.

gsmet commented:
@radcortez BTW, maybe I was unclear but the rest  is working. The two remaining issues are the ones I mentioned about the Redis extension.

gsmet committed:
Use new SmallRye Config map default feature

radcortez commented on extensions/redis-client/deployment/src/main/java/io/quarkus/redis/deployment/client/RedisClientProcessor.java:
```diff
@@ -294,7 +294,8 @@ private static List<String> getRedisLoadScript(RedisClientBuildTimeConfig config
             return List.of("import.redis");
         }
-        var scripts = config.loadScript;
+        // TODO @radcortez, this apparently returns a String now instead of a List<String>
+        var scripts = config.loadScript();
```
You need to map it like this:
`Optional<List<@WithConverter(TrimmedStringConverter.class)String>> loadScript();`

`@WithConverter` now applies in the exact nested element type.

gsmet replied:
That's what I did and the class cast error:

```java
    @ConfigDocDefault("import.redis in DEV, TEST ; no-file otherwise")
    Optional<List<@WithConverter(TrimmedStringConverter.class) String>> loadScript();
```

But it looks like the converter is somehow not applied:

```
Caused by: io.quarkus.runtime.configuration.ConfigurationException: Unable to find file referenced in 'quarkus.redis.redis-load-script=import/my-import.redis,  sample.redis'. Remove property or add file to your path.
	at io.quarkus.redis.deployment.client.RedisClientProcessor.preloadRedisData(RedisClientProcessor.java:263)
	at io.quarkus.redis.deployment.client.RedisClientProcessor.init(RedisClientProcessor.java:185)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```

radcortez replied:
It needs to be defined as:
`Optional<@WithConverter(TrimmedStringConverter.class) List<String>> loadScript();`

Sorry for misleading in the previous commented. Just pushed a fix.

gsmet replied:
Mkay. TBH, that's odd, I would have expected the converter to be placed on its target rather than the container.

radcortez replied:
https://github.com/smallrye/smallrye-config/pull/948

radcortez commented:
> @radcortez BTW, maybe I was unclear but the rest is working. The two remaining issues are the ones I mentioned about the Redis extension.

No, I've got it :)

I've already replied to both.

gsmet commented on extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/client/config/ProxyConfig.java:
```diff
@@ -3,41 +3,38 @@
      */
-    @ConfigItem
-    public String host;
+    Optional<String> host();
```
@cescoffier this is an important change. Before this change, you would have proxy enabled as long as you defined one of the property in `ProxyConfig` whatever it was.
Due to a change on how Optional groups are handled, I had to change it to the following behavior: the proxy is enabled if an `host` is defined. I think this is an acceptable trade-off.

gsmet replied:
@cescoffier just wanted to make sure you have seen ^ as I hope this PR will be ready really soon.

gsmet commented:
I'm back at this and I /think/ I fixed most of the issues now. I have one last one though that is extremely weird.

When I run `mvn clean install -f integration-tests/maven/ -Dit.test=DevMojoIT#testCapabilitiesMissing` (note that a few other Surefire tests will run before this one but it shouldn't be too long) after a full build, I get:

```
2023-07-25 12:32:02,876 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to recover after failed start: java.lang.RuntimeException: java.lang.ClassNotFoundException: io.quarkus.runtime.configuration.RuntimeOverrideConfigSource$$GeneratedMapHolder
	at io.quarkus.runtime.configuration.RuntimeOverrideConfigSource.<init>(RuntimeOverrideConfigSource.java:27)
	at io.quarkus.runtime.configuration.ConfigUtils.configBuilder(ConfigUtils.java:94)
	at io.quarkus.runtime.configuration.ConfigUtils.configBuilder(ConfigUtils.java:79)
	at io.quarkus.runtime.configuration.ConfigUtils.configBuilder(ConfigUtils.java:67)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder.startServerAfterFailedStart(VertxHttpRecorder.java:247)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup.handleFailedInitialStart(VertxHttpHotReplacementSetup.java:61)
	at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.startupFailed(RuntimeUpdatesProcessor.java:1150)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:151)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:423)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:55)
	at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:138)
	at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:93)
	at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:131)
	at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:62)
Caused by: java.lang.ClassNotFoundException: io.quarkus.runtime.configuration.RuntimeOverrideConfigSource$$GeneratedMapHolder
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:527)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:516)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:466)
	at io.quarkus.runtime.configuration.RuntimeOverrideConfigSource.<init>(RuntimeOverrideConfigSource.java:23)
	... 13 more

```

gsmet commented:
And ready \o/. Thanks for your help on this @radcortez !

geoand commented:
🏆

radcortez commented:
> And ready \o/. Thanks for your help on this @radcortez !

You did all the work. I've just fixed my own bugs :)

gsmet commented:
@radcortez unfortunately, I lost your commit when I rebased and I remember it was just a line but couldn't figure it out. Could you again push your fix on top of the rebased branch? Sorry and thanks!

radcortez committed:
Create empty config with default and discovered sources

radcortez commented:
Hum, now sure why but my commit is only showing in your branch and the PR did not update automatically.

gsmet commented:
That's very odd. Let me try to get the commit, rebase and force push.

gsmet commented:
It worked, let's see what CI has to say.

yrodiere commented on extensions/redis-client/deployment/src/test/java/io/quarkus/redis/client/deployment/RedisConfigClientNamesTest.java:
```diff
@@ -132,18 +98,90 @@ void testWithDisabledDefaultDevServiceWithHostsConfig() {
+        };
+
+        return devServiceConfiguration;
+    }
```
I'd suggest proper mocks here too.

yrodiere commented on extensions/datasource/runtime/src/main/java/io/quarkus/datasource/runtime/DataSourcesBuildTimeConfig.java:
```diff
@@ -55,22 +58,17 @@ public class DataSourcesBuildTimeConfig {
-        }
-        DataSourceBuildTimeConfig dataSourceBuildTimeConfig = namedDataSources.get(dataSourceName);
-        return Objects.requireNonNullElseGet(dataSourceBuildTimeConfig, DataSourceBuildTimeConfig::new);
+    default boolean hasNamedDataSources() {
+        return dataSources().keySet().size() > 1 || !dataSources().containsKey(DataSourceUtil.DEFAULT_DATASOURCE_NAME);
     }
```
_Ackshually_... this would return true for an empty config. Which might be fine, but I wouldn't take the risk?

```suggestion
    default boolean hasNamedDataSources() {
        return dataSources().keySet().size() > 1
                        || !dataSources().isEmpty()
                        && !dataSources().containsKey(DataSourceUtil.DEFAULT_DATASOURCE_NAME);
    }
```

yrodiere commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSourcesJdbcRuntimeConfig.java:
```diff
@@ -2,37 +2,48 @@
     /**
      * The default datasource.
      */
-    @ConfigItem
-    public DataSourceJdbcRuntimeConfig jdbc;
+    DataSourceJdbcRuntimeConfig jdbc();
 
     /**
      * Additional named datasources.
      */
     @ConfigDocSection
     @ConfigDocMapKey("datasource-name")
-    @ConfigItem(name = ConfigItem.PARENT)
-    public Map<String, DataSourceJdbcOuterNamedRuntimeConfig> namedDataSources;
+    @WithParentName
+    @WithDefaults
+    Map<String, DataSourceJdbcOuterNamedRuntimeConfig> namedDataSources();
```
Any reason not to merge these into a single `datasources()` method?

If so, can you please add a comment here so I don't break your stuff next time I end up in this part of the code? :)

yrodiere commented on extensions/datasource/deployment-spi/src/main/java/io/quarkus/datasource/deployment/spi/DevServicesDatasourceResultBuildItem.java:
```diff
@@ -3,34 +3,36 @@
-    public DevServicesDatasourceResultBuildItem(DbResult defaultDatasource, Map<String, DbResult> namedDatasources) {
-        this.defaultDatasource = defaultDatasource;
-        this.namedDatasources = Collections.unmodifiableMap(namedDatasources);
+    public DevServicesDatasourceResultBuildItem(Map<String, DbResult> namedDatasources) {
+        this.dataSources = Collections.unmodifiableMap(namedDatasources);
```
```suggestion
    public DevServicesDatasourceResultBuildItem(Map<String, DbResult> dataSources) {
        this.dataSources = Collections.unmodifiableMap(dataSources);
```

yrodiere commented on extensions/datasource/deployment-spi/src/main/java/io/quarkus/datasource/deployment/spi/DevServicesDatasourceProvider.java:
```diff
@@ -9,7 +9,7 @@
 
     RunningDevServicesDatasource startDatabase(Optional<String> username, Optional<String> password,
-            Optional<String> datasourceName,
+            String datasourceName,
```
This (and maybe other changes) will break backwards compatibility with some Quarkiverse extensions such as the one providing the SQLite JDBC driver.

Just mentioning in case you weren't aware, but I don't personally think it's a big issue.

gsmet replied:
Yeah. It might not be worth breaking it. I'll revisit it.

yrodiere commented on extensions/redis-client/deployment/src/test/java/io/quarkus/redis/deployment/client/RedisConfigClientNamesTest.java:
```diff
@@ -0,0 +1,187 @@
+            @Override
+            public boolean enabled() {
+                return enabled;
+            }
+
+            @Override
+            public Optional<String> imageName() {
+                return Optional.empty();
+            }
+
+            @Override
+            public OptionalInt port() {
+                return OptionalInt.empty();
+            }
+
+            @Override
+            public boolean shared() {
+                return false;
+            }
+
+            @Override
+            public String serviceName() {
+                return null;
+            }
+
+            @Override
+            public Map<String, String> containerEnv() {
+                return Map.of();
+            }
+        };
+
+        DevServiceConfiguration devServiceConfiguration = new DevServiceConfiguration() {
+
+            @Override
+            public DevServicesConfig devservices() {
+                return devServicesConfig;
+            }
+        };
+
+        return devServiceConfiguration;
```
Same here, I'd suggest mockito.

yrodiere reviewed (commented):
LGTM overall, though I'd be lying if I said this was a pleasant read :x 

Also, I added a few minor comments below.

gsmet commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSourceJdbcRuntimeConfig.java:
```diff
@@ -1,100 +1,95 @@
      */
-    @ConfigItem
-    public int minSize = 0;
+    @WithDefault("0")
+    int minSize();
```
I had some surprises about defaults and `@ConfigMapping`. I think we have to make them explicit now. At least I had an error message when I tried to not be at first.

gsmet commented on extensions/vertx/runtime/src/test/java/io/quarkus/vertx/core/runtime/VertxCoreProducerTest.java:
```diff
@@ -100,72 +168,357 @@ public void accept(VertxOptions vertxOptions) {
+                    return false;
+                }
+            };
+        }
+
+        @Override
+        public AddressResolverConfiguration resolver() {
+            return new AddressResolverConfiguration() {
+
+                @Override
+                public Duration queryTimeout() {
+                    return Duration.ofSeconds(5);
+                }
+
+                @Override
+                public int maxQueries() {
+                    return 4;
+                }
+
+                @Override
+                public int cacheNegativeTimeToLive() {
+                    return 0;
+                }
+
+                @Override
+                public int cacheMinTimeToLive() {
+                    return 0;
+                }
+
+                @Override
+                public int cacheMaxTimeToLive() {
+                    return Integer.MAX_VALUE;
+                }
+            };
+        }
+
+        @Override
+        public boolean preferNativeTransport() {
+            return false;
+        }
```
I thought about it and wasn't very excited at the idea either. If you have a strong opinion about it, feel free to change it in a follow-up PR.

gsmet commented on extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/config/EventBusConfiguration.java:
```diff
@@ -5,141 +5,129 @@
-    @ConfigItem
-    public int reconnectAttempts;
+    @WithDefault("0")
+    int reconnectAttempts();
```
See my answer above. For booleans, SmallRye Config was complaining so I made them explicit.

gsmet committed:
Improve compatibility for DevServicesDatasourceProvider
